features/environment.py

```
@@ -704,6 +684,49 @@ def create_instance_with_uat_installed(
     return inst
 
 
+def _get_user_data_for_instance(
+    config: UAClientBehaveConfig, series: str
+) -> str:
+    if config.install_from.value in (
+        InstallationSource.ARCHIVE.value,
+        InstallationSource.LOCAL.value,
+    ):
+        return ""
+
+    user_data = "#cloud-config\n"
+
+    if "pro" in config.machine_type:
+        user_data = USERDATA_BLOCK_AUTO_ATTACH_IMG
+
+    if config.userdata_file and os.path.exists(config.userdata_file):
+        with open(config.userdata_file, "r") as stream:
+            user_data = stream.read()
+
+    if config.install_from.value == InstallationSource.PROPOSED.value:
+        user_data += USERDATA_RUNCMD_ENABLE_PROPOSED.format(series=series)
+        return user_data
+
+    if config.install_from.value == InstallationSource.DAILY.value:
+        user_data += USERDATA_RUNCMD_PIN_PPA.format("daily")
+        ppa = UA_PPA_TEMPLATE.format("daily")
+        ppa_keyid = DEFAULT_UA_PPA_KEYID
+    elif config.install_from.value == InstallationSource.STAGING.value:
+        user_data += USERDATA_RUNCMD_PIN_PPA.format("staging")
+        ppa = UA_PPA_TEMPLATE.format("staging")
+        ppa_keyid = DEFAULT_UA_PPA_KEYID
+    elif config.install_from.value == InstallationSource.CUSTOM.value:
+        ppa = config.custom_ppa or ""
+        if ppa.startswith("ppa:"):
+            ppa = ppa.replace("ppa:", "http://ppa.launchpad.net/") + "/ubuntu"
+        ppa_keyid = config.custom_ppa_keyid or ""
```

Should we fail loudly here if any of the `ppa` variables are unset ?